### PR TITLE
Extract Tar bug fix

### PR DESCRIPTION
**Pull request by kelliejos**
Fix for [issue #833](https://github.com/MicrosoftDocs/vsts-docs/issues/830#event-1639300140) on the vsts-docs repo.

The JavaToolInstaller was failing for tar files on Linux machines.  This was due to a bug in how the files were being extracted.  Updated the logic to correctly extract the jdk from the tar file and set JAVA_HOME to the extract location.

### CODE REVIEW

**Comment by lkillgore** [comment]
Do we have a canary build for this?

**Comment by kelliejos** [comment]
I have created a canary build for this case as part of bug fix.  I verified that the extraction logic works for tar.gz files as well, without the -z flag.

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 1]
```diff
@@ -65,9 +65,13 @@ export class JavaFilesExtractor {
...
-        const tr: tr.ToolRunner = taskLib.tool('tar');
-        tr.arg(['xzC', destinationFolder, '-f', file]);
-        tr.exec();
+        const tarLocation = taskLib.which('tar', true);
+        const tar = taskLib.tool(tarLocation);
+        tar.arg('-xvf');
+        tar.arg(file);
+        tar.arg('-C');
+        tar.arg(destinationFolder);
+        return tar.execSync();
```
Comment (madhurig) [inline Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts]: Why making this synchronous and why adding the "return"?

Comment (madhurig) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts]: You can await on the tr.exec() too

Comment (kelliejos) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts]: That was a mistake. I'll remove the return.

**Comment by yacaovsnc** [comment]
Seems like GNU tar can auto recognize compression algorithms for a long time... -_-  .Z/.bz2/.gz should all work without their specific flags now.

**Comment by brcrista** [comment]
While you are making this change, could you have the task recognize `.tgz` as a tar extension?  It's just an abbreviation for `.tar.gz` but is fairly common

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 2]
```diff
@@ -65,9 +65,13 @@ export class JavaFilesExtractor {
 
     private async tarExtract(file, destinationFolder) {
```
Comment (brcrista) [inline Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts]: We should be using `extractTar` from vsts-task-tool-lib for this.  I see the implementations are the same.  If this broke, we should push the fix upstream

**Comment by brcrista** [comment]
While you are fixing this, I noticed some time ago that there is are a couple small bugs in `extractFiles` with the error handling.  Trying to stat a file that does not exist will throw rather than returning a null stats object.  Also, you need to check that the file is definitely a file, since there are other filesystem objects besides files and directories (at least on Unix-like systems).  To fix the bug, just import `fs` and use this code instead:

```
if (!fs.existsSync(compressedFile)) {
    throw new Error(tl.loc('ExtractNonExistFile', compressedFile));
} else if (!fs.statSync(compressedFile).isFile()) {
    throw new Error(tl.loc('ExtractNonFileFailed', compressedFile));
}
```

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts (hunk 2 above)
Comment (kelliejos) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts]: The extractTar method does not allow for specifying which destination folder should be used.  Using the extractTar method would require a change to both the vsts-tasks code base as well as the vsts-tasks-tool-lib code. In order to get this fix out sooner, I will be porting my current fix to the previous release and will abandon this PR in favor of making changes to both repos.

**Comment by kelliejos** [comment]
Abandoning this PR in favor of implementing this fix in the previous release branch.  I will update the vsts-task-tool-lib extractTar method to accept an optional destination directory parameter and then introduce a new PR here that will will use the tool-lib method.

**Pull request closed without merging**